File: src/main/java/info/szadkowski/contact/controller/MessageController.java
package info.szadkowski.contact.controller;

import info.szadkowski.contact.throttle.spring.aspect.Throttle;
import info.szadkowski.contact.model.MessageRequest;
import info.szadkowski.contact.service.MessageService;
import info.szadkowski.contact.template.TemplateFormatter;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Map;

@RestController
@RequestMapping(path = "/v1")
public class MessageController {
  private final MessageService messageService;
  private final TemplateFormatter formatter;

  public MessageController(MessageService messageService,
                           TemplateFormatter formatter) {
    this.messageService = messageService;
    this.formatter = formatter;
  }

  @Throttle
  @RequestMapping(path = "/message", consumes = MediaType.APPLICATION_JSON_UTF8_VALUE)
  public void sendMessage(@RequestBody Map<String, String> message) {
    messageService.send(MessageRequest.builder()
            .subject(message.get("subject"))
            .content(formatter.format(message))
            .build());
  }
}

File: src/main/kotlin/info/szadkowski/contact/controller/MessageController.kt
package info.szadkowski.contact.controller

import info.szadkowski.contact.exception.ThrottledRequestException
import info.szadkowski.contact.model.MessageRequest
import info.szadkowski.contact.service.MessageService
import info.szadkowski.contact.template.TemplateFormatter
import info.szadkowski.contact.throttle.Throttler
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.reactor.ReactorContext
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import kotlin.coroutines.coroutineContext

@RestController
@RequestMapping(path = ["/v1"])
class MessageController(
    private val messageService: MessageService,
    private val ipThrottler: Throttler,
    private val allThrottler: Throttler,
    private val format: TemplateFormatter
) {

    @ExperimentalCoroutinesApi
    @PostMapping(path = ["/message"], consumes = [MediaType.APPLICATION_JSON_VALUE])
    suspend fun sendMessage(@RequestBody message: Map<String, String>) {
        val ip = coroutineContext[ReactorContext]!!.context.get<String>("ip")
        if (!ipThrottler.canProcess(ip) || !allThrottler.canProcess("all")) {
            throw ThrottledRequestException()
        }
        messageService.send(
            MessageRequest(
                subject = message["subject"],
                content = message.format()
            )
        )
    }
}
